### ISSUE TITLE
Use separate processes instead of threads for paralellizing tasks

### ISSUE THREAD

**Issue by reidab** [frictionlessdata/goodtables-py#255]
I've recently been setting up a nightly task that uses goodtables to validate a data package with 15 tables including one with about 1.5M lines and nine others with over 20k. I'm glad to see goodtables churning through them all in parallel, but would love it not to be constrained to a single CPU core.

**Comment by vitorbaptista** [comment]
Good point, @reidab. I agree on the change, and it should be limited to https://github.com/frictionlessdata/goodtables-py/blob/d1cac3bb525f00edeccc73acdb8a2693c86488cf/goodtables/inspector.py. If you have some time, this would make a great contribution.

**Comment by roll** [comment]
@reidab 
Have you got some initial profiling results of #256? It's interesting. 

At first place, it uses threads because goodtables is very IO-intensive so there is a hypothesis hat CPU is not really relevant. But some `tableschema-py` operations are not very effective so probably multiprocessing could speed it up.

**Comment by reidab** [comment]
@roll For my specific use case, there's definitely a speed boost. 

I'm seeing the validation process taking 100% of one core through its entire run, which leads me to believe that something is CPU-bound. 

I'm validating a data package with table schemas against the nightly CSV dumps from https://ridb.recreation.gov, which includes several 100k+ row tables (see below). I'm particularly interested in finding all the places where the input data has things like invalid URIs or email addresses, so I'm specifying string formats on several fields.

```
reidab ➜ wc -l data/*                                                                                                                                                                                         
      43 data/Activities_API_v1.csv
 1536734 data/Attributes_API_v1.csv
  100111 data/Campsites_API_v1.csv
  130956 data/EntityActivities_API_v1.csv
       3 data/Events_API_v1.csv
   24771 data/Facilities_API_v1.csv
   25594 data/FacilityAddresses_API_v1.csv
   50928 data/Links_API_v1.csv
   19399 data/Media_API_v1.csv
      28 data/MemberTours_API_v1.csv
   24623 data/OrgEntities_API_v1.csv
      31 data/Organizations_API_v1.csv
     401 data/PermitEntranceZones_API_v1.csv
     970 data/PermitEntrances_API_v1.csv
  543441 data/PermittedEquipment_API_v1.csv
    4120 data/RecAreaAddresses_API_v1.csv
   19706 data/RecAreaFacilities_API_v1.csv
   24699 data/RecAreas_API_v1.csv
     274 data/Tours_API_v1.csv
 2506832 total
```

Some quick benchmarks using `time` on a 2012 MacBook Pro i7, 8 core:

1. Running from master on the whole data set takes over 13 minutes and uses one core:

       (master)$ time goodtables datapackage.json
       783.00s user 18.57s system 101% cpu 13:12.47 total

2. With #256 applied, the whole set takes 11.5 minutes. Watching a CPU graph, I could see it using all 8 cores at the beginning, but dropping down to just one once (presumably) the only task left was the 1.5M line CSV:

       (#256)$ time goodtables datapackage.json
       1892.50s user 19.29s system 273% cpu 11:38.87 total

3. When a row limit is applied each file gets a much closer to constant processing time, so the long tail of one-core time in (2) doesn't happen and we get a more dramatic improvement from 11s down to 3.5s:

       (master)$ time goodtables --row-limit=1000 datapackage.json
       10.43s user 1.38s system 104% cpu 11.324 total
       ($256)$ time goodtables --row-limit=1000 datapackage.json
       16.57s user 0.58s system 498% cpu 3.439 total

**Comment by roll** [comment]
@reidab 
Thanks. It's great. All these string checks are definitely CPU-intensive.

Please take a look at my comment in the PR about the table schema serialization if it could help to simplify #256.

**Comment by roll** [comment]
@reidab
#256 is closed due to inactivity but of course please feel free to re-open if you have time to finish it

Anyone who also is interested in working on this feature please take into account the work done in #256

**Comment by rudiharms** [comment]
If you can somehow get the tables into seperate files, you can multi process it on seperate cores.

see code below:



from goodtables import validate
import sys
import multiprocessing 
#import time
#multithreading of validations by passing in comma seperated list of files in argument 1.
#Example "@commaseperatedMultipleFiles_String" "@Schema" 500000


def multiprocessing_func(filetoValidate,schema,rowsperfile,fileNum,rowlimit):
    try:       
        #report = validate(r'C:\temp\tempo\CMFTypeCC_5_2019mmed.csv', schema=r'\\cmssql01\programmersHeaven\Returns\DDDR\schemas\JsonDDDRPartYearReturnTypeCC5_2019_8.json',error_limit=5,row_limit=10000000,checks=['schema'])
        report = validate(source=filetoValidate, schema=schema, error_limit=3,row_limit=100000000,checks=['schema']) # 
        reportString = ''

        if report['error-count'] > 0: #if error count greater than 0, package loaded fine                       
            if report['valid'] == 0:                 
                #print('Errors : ' + str(report['error-count'])) # 3
                  
                lengthToPrint = 300
                for i in (report['tables'][0]['errors']):
                    if len(i['message']) < 300:
                        lengthToPrint = len(i['message'])

                    reportString = ''
                    reportString = i['message'][:lengthToPrint] + '.'
                    #retrieve all numbers from the list
                    intList = [int(s) for s in reportString.split() if s.isdigit()]
                    for i in intList: 
                        findstr = 'row ' + str(i) + ' '
                        replacestr = 'row ' + str((int(fileNum) * int(rowsperfile)) + int(i) - int(fileNum)) + ' '
                        reportString = reportString.replace(findstr,replacestr)
                    print(reportString)                  
        else:
            if len(report['warnings']) > 0:# if warnings exist, something failed with the load
                print(report['warnings'][0])
            else:
                if report['valid'] == 1:
                    r=""#fake blank message
                    #print('')
    except:
        print("Unexpected error:", sys.exc_info()[0])
try:
  if __name__ == '__main__':
    #starttime = time.time()
    processes = []
    if len(sys.argv) > 1:
        list1 = sys.argv[1].split(',')       
        schema = sys.argv[2]
        rowsperfile = int(sys.argv[3])
        rowlimit = int(50/len(list1))#set max errors to 50 for file. Set error limit per file
        for i in range(0,len(list1)):#start muliple threads.can send in different files in the argument
            filetoValidate = list1[i]           
            p = multiprocessing.Process(target=multiprocessing_func, args=(filetoValidate,schema,rowsperfile,i,rowlimit))
            processes.append(p)
            p.start()
        
        for process in processes:
            process.join()#blocks calling thread until joined processes are complete.
    else:    
            print('Argmuments not passed in Error.')   
        
    #print('That took {} seconds'.format(time.time() - starttime))
except:
    print("Unexpected error:", sys.exc_info()[0])

**Comment by roll** [comment]
MERGED to #341
